### ISSUE TITLE
Building a composite mask with attention_mask from tokenizers

### ISSUE THREAD

**Issue by lhallee** [pytorch-labs/attention-gym#98]
Trying to compare if document masking is any faster than utilizing the HF transformers masking convention. Really confused why the examples below are not equivalent, I am probably constructing the mask_mods wrong. Would appreciate any help.

```python
import torch
from torch.nn.attention.flex_attention import create_block_mask, flex_attention
from transformers import EsmTokenizer

torch.set_default_device("cuda")
torch.manual_seed(0)
torch._dynamo.config.cache_size_limit = 1000
flex_attention = torch.compile(flex_attention, dynamic=False)

tokenizer = EsmTokenizer.from_pretrained("facebook/esm2_t33_650M_UR50D")


heads = 8
hidden_dim = 512
sliding_window_size = 64
seq_a = 'MYDSNIFEKVNQYKFLYIWWLIMINVNH'
seq_b = 'VPAALSAHAVLVDHHDQLLAALRVLLPVRHWEGAVVTPEQTRRQHVAQRDMVVRSLLLAPTPEERGEERPRKPRQVTVVPVVSREPGAHAYRLPIVREARYL'
batch = [seq_a, seq_b, seq_a, seq_b]


### Example 1 with (1, seq_len * batch_size)
input_ids = []
for seq in batch:
    input_ids.extend(tokenizer(seq, add_special_tokens=False, truncation=True, max_length=1024, padding=False).input_ids)

input_ids = torch.tensor(input_ids).flatten()

def doc_mask_mod(b, h, q_idx, kv_idx): # BERT style document mask with sliding window
    bidirectional_sliding_window_mask = torch.abs(q_idx - kv_idx) < sliding_window_size
    doc_mask = docs[q_idx] == docs[kv_idx]
    return bidirectional_sliding_window_mask & doc_mask

input_ids = input_ids.flatten()
docs = (input_ids == tokenizer.cls_token_id).cumsum(dim=0)  # shape: [S]
S = len(input_ids)
block_mask = create_block_mask(doc_mask_mod, None, None, S, S)

# BHLD
q = torch.randn(1, heads, S, hidden_dim // heads).cuda()
k = torch.randn(1, heads, S, hidden_dim // heads).cuda()
v = torch.randn(1, heads, S, hidden_dim // heads).cuda()

att_1 = flex_attention(q, k, v, block_mask=block_mask).detach().cpu()
print(att_1.shape) # (1, 8, S, 64)


### Example 2 with more traditional (batch_size, seq_len)
tokenized = tokenizer(batch, return_tensors='pt', padding=True, add_special_tokens=False, truncation=True, max_length=1024)
input_ids = tokenized.input_ids.cuda() # (B, S)
attention_mask = tokenized.attention_mask.bool().cuda() # (B, S)

def batched_mask_mod(b, h, q_idx, kv_idx): # BERT style document mask based on attention mask with sliding window
    bidirectional_sliding_window_mask = torch.abs(q_idx - kv_idx) < sliding_window_size
    doc_mask = attention_mask[b, q_idx] & attention_mask[b, kv_idx]
    return bidirectional_sliding_window_mask & doc_mask

B = input_ids.shape[0]
S = input_ids.shape[1]

q = torch.randn(B, heads, S, hidden_dim // heads).cuda()
k = torch.randn(B, heads, S, hidden_dim // heads).cuda()
v = torch.randn(B, heads, S, hidden_dim // heads).cuda()

block_mask = create_block_mask(batched_mask_mod, None, None, S, S)

att_2 = flex_attention(q, k, v, block_mask=block_mask).detach().cpu()
print(att_2.shape) # (4, 8, S, 64)

# The first len(seq_a) of each should be the attention output of the first sequence
close = att_1[0, :, :len(seq_a), :].isclose(att_2[0, :, :len(seq_a), :])
print(close.any()) # False, expecting True
```

**Comment by drisspg** [comment]
Me and fable did a few rounds on this and I think we found 3 bugs:

```Py
"""Corrected repro for attention-gym #98: packed vs padded BERT-style masking.

The original script had three bugs (marked FIX 1/2/3 below):
  1. Different random q/k/v were drawn for the packed and padded layouts, so the
     outputs could never match regardless of masking.
  2. `docs` was built from CLS-token positions, but tokenization used
     `add_special_tokens=False`, so no CLS tokens exist -> `docs` was all zeros and
     the sliding window leaked across sequence boundaries in the packed layout.
  3. `create_block_mask(..., B=None, ...)` with a batch-dependent mask_mod evaluates
     the mask only for b=0 and broadcasts it across the batch; it must get `B=B`.
"""

import torch
from torch.nn.attention.flex_attention import create_block_mask, flex_attention
from transformers import EsmTokenizer

torch.set_default_device("cuda")
torch.manual_seed(0)
flex_attention = torch.compile(flex_attention, dynamic=False)

tokenizer = EsmTokenizer.from_pretrained("facebook/esm2_t33_650M_UR50D")

heads = 8
hidden_dim = 512
head_dim = hidden_dim // heads
sliding_window_size = 64
seq_a = "MYDSNIFEKVNQYKFLYIWWLIMINVNH"
seq_b = "VPAALSAHAVLVDHHDQLLAALRVLLPVRHWEGAVVTPEQTRRQHVAQRDMVVRSLLLAPTPEERGEERPRKPRQVTVVPVVSREPGAHAYRLPIVREARYL"
batch = [seq_a, seq_b, seq_a, seq_b]

lengths = [
    len(tokenizer(s, add_special_tokens=False, truncation=True, max_length=1024).input_ids)
    for s in batch
]

# FIX 1: draw q/k/v once per sequence, then assemble BOTH layouts from the same
# tensors. The original drew independent randn for each layout.
per_seq_qkv = [torch.randn(3, heads, length, head_dim) for length in lengths]


### Example 1: packed layout (1, sum(lengths))

# FIX 2: derive document ids from sequence lengths. The original used
# `(input_ids == cls_token_id).cumsum(0)`, but with add_special_tokens=False there
# are no CLS tokens, so every token landed in document 0 and the window crossed
# sequence boundaries.
docs = torch.repeat_interleave(torch.arange(len(batch)), torch.tensor(lengths))


def doc_mask_mod(b, h, q_idx, kv_idx):
    window = (q_idx - kv_idx).abs() < sliding_window_size
    return window & (docs[q_idx] == docs[kv_idx])


S_packed = sum(lengths)
q, k, v = torch.cat(per_seq_qkv, dim=2).unsqueeze(1)
block_mask = create_block_mask(doc_mask_mod, None, None, S_packed, S_packed)
att_packed = flex_attention(q, k, v, block_mask=block_mask)
print("packed:", att_packed.shape)


### Example 2: padded layout (B, max_len)

B, S_padded = len(batch), max(lengths)
attention_mask = torch.zeros(B, S_padded, dtype=torch.bool)
padded_qkv = torch.zeros(3, B, heads, S_padded, head_dim)
for b, (length, qkv) in enumerate(zip(lengths, per_seq_qkv)):
    attention_mask[b, :length] = True
    padded_qkv[:, b, :, :length] = qkv


def batched_mask_mod(b, h, q_idx, kv_idx):
    window = (q_idx - kv_idx).abs() < sliding_window_size
    return window & attention_mask[b, q_idx] & attention_mask[b, kv_idx]


# FIX 3: pass B=B. With B=None, create_block_mask evaluates the mask_mod only at
# b=0 (seq_a's padding pattern) and broadcasts it to the whole batch, silently
# giving every row seq_a's mask.
block_mask = create_block_mask(batched_mask_mod, B, None, S_padded, S_padded)
att_padded = flex_attention(*padded_qkv, block_mask=block_mask)
print("padded:", att_padded.shape)


offsets = [0, *torch.tensor(lengths).cumsum(0).tolist()]
for b, length in enumerate(lengths):
    torch.testing.assert_close(
        att_packed[0, :, offsets[b] : offsets[b] + length],
        att_padded[b, :, :length],
        rtol=1e-4,
        atol=1e-5,
    )
print("all sequences match between packed and padded layouts")

```

From your orginal script; with this they are all equal
